File: KCD/HMNyukyoStartCommand.m
//
//  HMNyukyoStartCommand.m
//  KCD
//
//  Created by Hori,Masaki on 2014/09/03.
//  Copyright (c) 2014年 Hori,Masaki. All rights reserved.
//

#import "HMNyukyoStartCommand.h"

#import "HMServerDataStore.h"
#import "HMKCShipObject+Extensions.h"
#import "HMKCMaterial.h"


@implementation HMNyukyoStartCommand
+ (void)load
{
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		[HMJSONCommand registerClass:self];
	});
}

+ (BOOL)canExcuteAPI:(NSString *)api
{
	if([api isEqualToString:@"/kcsapi/api_req_nyukyo/start"]) return YES;
	
	return NO;
}
- (void)execute
{
	id obj = self.arguments[@"api_highspeed"];
	if(!obj) return;
	if(![obj boolValue]) return;
	
	
	HMServerDataStore *store = [HMServerDataStore oneTimeEditor];
	
    NSError *error = nil;

	NSString *shipId = self.arguments[@"api_ship_id"];
    NSArray<HMKCShipObject *> *ships = [store objectsWithEntityName:@"Ship"
															  error:&error
													predicateFormat:@"id = %@", @([shipId integerValue])];
	if(ships.count == 0) {
		if(error) {
			NSLog(@"Error: at %@ : %@", NSStringFromClass([self class]), error);
		}
		return;
	}
	ships[0].nowhp = ships[0].maxhp;
    
    NSArray<HMKCMaterial *> *materials = [store objectsWithEntityName:@"Material"
                                            predicate:nil
                                                error:&error];
    if(materials.count == 0) {
        if(error) {
            NSLog(@"Error: at %@ : %@", NSStringFromClass([self class]), error);
        }
        return;
    }
    NSNumber *bukkets = materials[0].kousokushuhuku;
    materials[0].kousokushuhuku = @(bukkets.integerValue - 1);
}
@end

File: KCD/HMNyukyoStartCommand.swift
//
//  HMNyukyoStartCommand.swift
//  KCD
//
//  Created by Hori,Masaki on 2015/01/17.
//  Copyright (c) 2015年 Hori,Masaki. All rights reserved.
//

import Cocoa

/// 高速修復材を併用したときにHPを回復させる
class HMNyukyoStartCommand: HMJSONCommand {
	required override init() {}
	
	override class func load() {
		// Mavericksにおいてload()のタイミングでNSArrayを生成することが出来ないため、遅延させる
		dispatch_after(dispatch_time(DISPATCH_TIME_NOW, Int64(1 * Double(NSEC_PER_MSEC))), dispatch_get_main_queue()) {
			HMJSONCommand.registerInstance(self())
		}
	}
	override class func canExcuteAPI(api: String) -> Bool {
		return api == "/kcsapi/api_req_nyukyo/start"
	}
	
	override func execute() {
		let obj = arguments["api_highspeed"] as String?
		if obj == nil { return }
		if let b = obj!.toInt() {
			if b == 0 { return }
		} else {
			return
		}
		
		if let shipID = arguments["api_ship_id"] as? String {
			var error: NSError? = nil
			let store = HMServerDataStore.oneTimeEditor()
			let predicate = NSPredicate(format: "id = \(shipID)")
			let array = store.objectsWithEntityName("Ship",
				sortDescriptors: nil,
				predicate: predicate,
				error: &error)
			if array.count == 0 {
				log("Could not find Ship id \(shipID)", argList: getVaList([]))
				return
			}
			
			if let ship = array[0] as? HMKCShipObject {
				ship.nowhp = ship.maxhp
			}
		}
	}
}
